File: frontend/src/types/session.ts
/**
 * Session domain types — mirror the session-service DTOs
 * (SessionResponse, BookingResponse, CalendarResponse, MeetingResponse,
 * BookingRequest, RescheduleRequestDto, CancellationRequest).
 */

export type SessionStatus =
  | 'SCHEDULED'
  | 'PENDING_APPROVAL'
  | 'APPROVED'
  | 'CONFIRMED'
  | 'PENDING'
  | 'IN_PROGRESS'
  | 'COMPLETED'
  | 'CANCELLED'
  | 'REJECTED'
  | 'RESCHEDULED'
  | 'NO_SHOW'
  | 'EXPIRED';

export type BookingStatus =
  | 'PENDING'
  | 'APPROVED'
  | 'REJECTED'
  | 'CANCELLED'
  | 'EXPIRED';

export interface MeetingInfo {
  id?: string;
  provider?: string;
  meetingId?: string;
  meetingUrl?: string;
  joinUrl?: string;
  startUrl?: string;
  password?: string;
  active?: boolean;
}

export interface Session {
  id: string;
  title?: string;
  description?: string;
  mentorId: string;
  learnerId: string;
  mentorName?: string;
  learnerName?: string;
  startTime?: string;
  endTime?: string;
  durationMinutes: number;
  timezone?: string;
  status: SessionStatus;
  topic?: string;
  category?: string;
  price?: number;
  currency?: string;
  free?: boolean;
  /** True for community mentoring sessions. */
  community?: boolean;
  /** Number of participants (community sessions). */
  participantCount?: number;
  /** Learner capacity of a community session (1–20). */
  maxParticipants?: number;
  /** Learners currently joined (excludes the host). */
  learnerCount?: number;
  /** Seats left before the session is full (0 = full). */
  remainingSeats?: number;
  recordingUrl?: string;
  notes?: string;
  outcome?: string;
  rating?: number;
  feedback?: string;
  cancellationReason?: string;
  rescheduleCount?: number;
  completedAt?: string;
  startedAt?: string;
  endedAt?: string;
  meetingLink?: MeetingInfo;
  /** When the join window opens (start minus the join window), per the backend. */
  joinAvailableAt?: string;
  /** Backend-computed: may the current user join right now? */
  joinAllowed?: boolean;
  /** Backend-computed: is a meeting link stored for this session? */
  sessionLinkAvailable?: boolean;
  createdAt?: string;
  updatedAt?: string;
}

/* ---------------- Request payloads (mirror backend) ---------------- */

export interface BookingRequest {
  mentorId: string;
  learnerId?: string;
  mentorName?: string;
  learnerName?: string;
  /** Learner email — used for booking status notifications (stored with the booking). */
  learnerEmail?: string;
  topic: string;
  description?: string;
  preferredDate?: string;
  preferredStartTime: string;
  preferredEndTime: string;
  durationMinutes: number;
  /** Selected mentor pricing plan — the backend computes the real cost. */
  pricingId?: string;
  /** Session cost in credits (1 credit = 10 minutes) — informational only. */
  credits?: number;
  timezone?: string;
  learnerMessage?: string;
}

export interface CommunitySessionRequest {
  topic: string;
  description?: string;
  startTime: string;
  endTime: string;
  timezone?: string;
  /** Cost in credits (0–3). 0 = TRUE FREE session. */
  cost: number;
  /** Max learners (1–20). */
  maxParticipants: number;
}

export interface CommunityAllowance {
  limit: number;
  used: number;
  remaining: number;
  month: string;
}

export interface CommunityImpact {
  completedSessions: number;
  learnersHelped: number;
  communityHours: number;
  level: number;
  levelLabel?: string | null;
  nextLevelAt: number;
  nextLevelLabel?: string | null;
}

export interface BookingResponse {
  id: string;
  mentorId: string;
  learnerId: string;
  sessionId?: string;
  mentorName?: string;
  learnerName?: string;
  learnerEmail?: string;
  topic?: string;
  description?: string;
  preferredDate?: string;
  preferredStartTime?: string;
  preferredEndTime?: string;
  durationMinutes?: number;
  /** Session cost in credits (0 for free/community). */
  price?: number;
  timezone?: string;
  status: BookingStatus;
  mentorMessage?: string;
  learnerMessage?: string;
  rejectionReason?: string;
  approvedAt?: string;
  rejectedAt?: string;
  expiresAt?: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface RescheduleRequest {
  sessionId: string;
  proposedStartTime: string;
  proposedEndTime: string;
  reason?: string;
}

export interface CancellationRequest {
  sessionId: string;
  reason?: string;
  cancellationType?: 'VOLUNTARY' | 'INVOLUNTARY' | 'NO_SHOW';
}

export interface SessionFilters {
  status?: SessionStatus;
  query?: string;
  page?: number;
  size?: number;
}

File: backend/session-service/src/main/java/com/skillinfinity/session/entity/Session.java
package com.skillinfinity.session.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "sessions")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Session {

    @Id
    @Column(name = "id", updatable = false, nullable = false)
    private UUID id;

    @Column(name = "title", nullable = false, length = 255)
    private String title;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "mentor_id", nullable = false)
    private UUID mentorId;

    @Column(name = "learner_id", nullable = false)
    private UUID learnerId;

    @Column(name = "mentor_name", length = 255)
    private String mentorName;

    @Column(name = "learner_name", length = 255)
    private String learnerName;

    @Column(name = "start_time", nullable = false)
    private LocalDateTime startTime;

    @Column(name = "end_time", nullable = false)
    private LocalDateTime endTime;

    @Column(name = "duration_minutes", nullable = false)
    private int durationMinutes;

    @Column(name = "timezone", length = 50)
    private String timezone;

    @Column(name = "status", nullable = false, length = 30)
    @Enumerated(EnumType.STRING)
    private com.skillinfinity.session.enumeration.SessionStatus status;

    @Column(name = "topic", length = 255)
    private String topic;

    @Column(name = "category", length = 100)
    private String category;

    @Column(name = "price", nullable = false)
    private double price;

    @Column(name = "currency", length = 3)
    private String currency;

    @Column(name = "is_free", nullable = false)
    private boolean free;

    /**
     * True for free community mentoring sessions (spec: 3 per learner per
     * month). Community sessions cost 0 credits and contribute to the
     * mentor's community recognition rather than the credit economy.
     */
    @Column(name = "is_community", nullable = false)
    @Builder.Default
    private boolean community = false;

    @Column(name = "recording_url", length = 500)
    private String recordingUrl;

    @Column(name = "notes", columnDefinition = "TEXT")
    private String notes;

    @Column(name = "outcome", columnDefinition = "TEXT")
    private String outcome;

    @Column(name = "rating")
    private Integer rating;

    @Column(name = "feedback", columnDefinition = "TEXT")
    private String feedback;

    @Column(name = "cancellation_reason", columnDefinition = "TEXT")
    private String cancellationReason;

    @Column(name = "cancelled_by")
    private UUID cancelledBy;

    @Column(name = "cancelled_at")
    private LocalDateTime cancelledAt;

    @Column(name = "completed_at")
    private LocalDateTime completedAt;

    @Column(name = "started_at")
    private LocalDateTime startedAt;

    @Column(name = "ended_at")
    private LocalDateTime endedAt;

    @Column(name = "booking_id")
    private UUID bookingId;

    @Column(name = "reschedule_count", nullable = false)
    private int rescheduleCount;

    @Column(name = "version")
    @Version
    private Long version;

    // Audit fields
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @Column(name = "created_by", updatable = false)
    private String createdBy;

    @Column(name = "updated_by")
    private String updatedBy;

    @PrePersist
    protected void onCreate() {
        if (id == null) {
            id = UUID.randomUUID();
        }
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}
